scale-encode/src/impls/composite.rs

```diff
@@ -35,119 +97,168 @@ use scale_info::{PortableRegistry, TypeDef};
 {
-    fn encode_as_type_to(
+    /// Construct a new [`Composite`] type by providing an iterator over
+    /// the fields that it contains.
+    ///
+    /// ```rust
+    /// use scale_encode::{ Composite, CompositeField };
+    /// use scale_info::PortableRegistry;
+    ///
+    /// Composite::<PortableRegistry, _>::new([
+    ///     (Some("foo"), CompositeField::new(&123)),
+    ///     (Some("bar"), CompositeField::new(&"hello"))
+    /// ].into_iter());
+    /// ```
+    pub fn new(vals: Vals) -> Self {
+        Composite {
+            vals,
+            marker: core::marker::PhantomData,
+        }
+    }
+
+    /// A shortcut for [`Self::encode_composite_as_type_to()`] for when you just
+    /// want it to allocate and return the encoded bytes.
+    pub fn encode_composite_as_type(
+        &self,
+        type_id: &R::TypeId,
+        types: &R,
+    ) -> Result<Vec<u8>, Error> {
+        let mut out = Vec::new();
+        self.encode_composite_as_type_to(type_id, types, &mut out)?;
+        Ok(out)
+    }
+
+    /// Encode this composite value as the provided type to the output bytes.
+    pub fn encode_composite_as_type_to(
         &self,
-        type_id: u32,
-        types: &PortableRegistry,
+        type_id: &R::TypeId,
+        types: &R,
         out: &mut Vec<u8>,
     ) -> Result<(), Error> {
-        let mut vals_iter = self.0.clone();
+        let vals_iter = self.vals.clone();
         let vals_iter_len = vals_iter.len();
 
         // Skip through any single field composites/tuples without names. If there
         // are names, we may want to line up input field(s) on them.
         let type_id = skip_through_single_unnamed_fields(type_id, types);
 
-        let ty = types
-            .resolve(type_id)
-            .ok_or_else(|| Error::new(ErrorKind::TypeNotFound(type_id)))?;
-
-        match &ty.type_def {
-            // If we see a tuple type, it'll have more than one field else it'd have been skipped above.
-            TypeDef::Tuple(tuple) => {
-                // If there is exactly one val, it won't line up with the tuple then, so
-                // try encoding one level in instead.
-                if vals_iter_len == 1 {
-                    return vals_iter
-                        .next()
-                        .unwrap()
-                        .1
-                        .encode_as_type_to(type_id, types, out);
-                }
-
-                let mut fields = tuple.fields.iter().map(|f| Field::unnamed(f.id));
-                self.encode_as_fields_to(&mut fields, types, out)
+        let v = visitor::new((out, vals_iter), |(out, mut vals_iter), _| {
+            // Rather than immediately giving up, we should at least see whether
+            // we can skip one level in to our value and encode that.
+            if vals_iter_len == 1 {
+                return vals_iter
+                    .next()
+                    .unwrap()
```

Comment:
I prefer expect over unwrap but this is clear to me with `vals_iter_len == 1` above